File: python/day3/main.py
from pathlib import Path
from argparse import ArgumentParser
from functools import reduce


def parse(file: Path):
    with file.open() as fp:
        for line in fp:
            yield [int(c) for c in line.strip()]


def pick_max(line, dl):
    def step(state, i):
        si, num = state
        ei = len(line) - (dl - i - 1)
        f = max(line[si:ei])
        return (line.index(f, si) + 1, num * 10 + f)
    return reduce(step, range(dl), (0, 0))[1]


def solve1(data, dl=2):
    return sum(pick_max(line, 2) for line in data)


def solve2(data):
    return sum(pick_max(line, 12) for line in data)
    

parser = ArgumentParser()
parser.add_argument("input", type=Path)
args = parser.parse_args()

nums = list(parse(args.input))
print(f"Day3 part1: {solve1(nums)}")
print(f"Day3 part2: {solve2(nums)}")

File: rust/day3/src/main.rs
use common::{get_lines, Result};
use std::env;
use std::path::Path;

fn parse(file: &Path) -> Result<Vec<Vec<u8>>> {
  let lines = get_lines(file)?;
  Ok(
    lines
      .iter()
      .map(|line| {
        line
          .chars()
          .map(|c| c.to_digit(10).unwrap() as u8)
          .collect()
      })
      .collect(),
  )
}

fn pick_max(line: &[u8], dl: usize) -> u64 {
  (0..dl)
    .fold((0usize, 0u64), |(si, num), i| {
      let ei = line.len() - (dl - i - 1);
      let f = *line[si..ei].iter().max().unwrap();
      let new_si = line[si..].iter().position(|&x| x == f).unwrap() + si + 1;
      (new_si, num * 10 + f as u64)
    })
    .1
}

fn solve1(data: &[Vec<u8>]) -> u64 {
  data.iter().map(|line| pick_max(line, 2)).sum()
}

fn solve2(data: &[Vec<u8>]) -> u64 {
  data.iter().map(|line| pick_max(line, 12)).sum()
}

fn main() -> Result<()> {
  let input = env::args().nth(1).expect("usage: day3 <input-file>");
  let nums = parse(Path::new(&input))?;
  println!("Day3 part1: {}", solve1(&nums));
  println!("Day3 part2: {}", solve2(&nums));
  Ok(())
}
